src/tools/database/elasticsearch.ts

```diff
@@ -0,0 +1,198 @@
+  JSONToolOutput,
+} from "@/tools/base.js";
+import { Cache } from "@/cache/decoratorCache.js";
+import { z } from "zod";
+import { Client, ClientOptions } from "@elastic/elasticsearch";
+import {
+  CatIndicesResponse,
+  IndicesGetMappingResponse,
+  SearchRequest,
+  SearchResponse,
+  SearchHit,
+} from "@elastic/elasticsearch/lib/api/types.js";
+import { ValidationError } from "ajv";
+
+type ToolRunOptions = BaseToolRunOptions;
+
+export interface ElasticSearchToolOptions extends BaseToolOptions {
+  connection: ClientOptions;
+}
+
+export class ElasticSearchTool extends Tool<
+  JSONToolOutput<any>,
+  ElasticSearchToolOptions,
+  ToolRunOptions
+> {
+  name = "ElasticSearchTool";
+
+  description = `Can query data from an ElasticSearch database. IMPORTANT: strictly follow this order of actions:
+   1. LIST_INDICES - retrieve a list of available indices
+   2. GET_INDEX_DETAILS - get details of index fields
+   3. SEARCH - perform search or aggregation query on a specific index or pass the original user query without modifications if it's a valid JSON ElasticSearch query`;
+
+  inputSchema() {
+    return z.object({
+      action: z
+        .enum(["LIST_INDICES", "GET_INDEX_DETAILS", "SEARCH"])
+        .describe(
+          "The action to perform. LIST_INDICES lists all indices, GET_INDEX_DETAILS fetches details for a specified index, and SEARCH executes a search or aggregation query",
+        ),
+      indexName: z
+        .string()
+        .optional()
+        .describe("The name of the index to query, required for GET_INDEX_DETAILS and SEARCH"),
+      query: z
+        .string()
+        .optional()
+        .describe("Valid ElasticSearch JSON search or aggregation query for SEARCH action"),
+      start: z.coerce
+        .number()
+        .int()
+        .min(0)
+        .default(0)
+        .optional()
+        .describe(
+          "The record index from which the query will start. Increase by the size of the query to get the next page of results",
+        ),
+      size: z.coerce
+        .number()
+        .int()
+        .min(0)
+        .max(10)
+        .default(10)
+        .optional()
+        .describe("How many records will be retrieved from the ElasticSearch query. Maximum is 10"),
+    });
+  }
+
+  static {
+    this.register();
+  }
+
+  public constructor(options: ElasticSearchToolOptions) {
+    super(options);
+    if (!options.connection.cloud && !options.connection.node && !options.connection.nodes) {
+      throw new ValidationError([
+        {
+          message: "At least one of the properties must be provided",
+          propertyName: "connection.cloud, connection.node, connection.nodes",
+        },
+      ]);
+    }
+  }
+
+  @Cache()
+  protected get client(): Client {
+    try {
+      return new Client(this.options.connection);
+    } catch (error) {
+      throw new ToolError(`Unable to connect to ElasticSearch: ${error}`, [], {
+        isRetryable: false,
+        isFatal: true,
+      });
```

Comment:
Do the `Client` instances immediately connect to the database? I don't think so.

Reply:
I could add client.ping() but it requires monitor or manage cluster privilege

Reply:
Another reason for not requiring ping:

[https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/client-connecting.html](https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/client-connecting.html
)

_Since v7.14.0, the client performs a required product check before the first call. This pre-flight product check allows the client to establish the version of Elasticsearch that it is communicating with. The product check requires one additional HTTP request to be sent to the server as part of the request pipeline before the main API call is sent. In most cases, this will succeed during the very first API call that the client sends. Once the product check completes, no further product check HTTP requests are sent for subsequent API calls._

Reply:
I will use client.info() but it requires the same privilege as above. I get more meaningful error message with this rather than client.ping()